drivers/counter/counter_gecko_rtcc.c

```diff
@@ -0,0 +1,354 @@
+		rtccCompMatchOutActionPulse,/* Don't care */
+		rtccPRSCh0,                 /* PRS is not used */
+		rtccInEdgeNone,             /* Capture input is not used */
+		rtccCompBaseCnt,            /* Compare with base CNT register */
+		0,                          /* Compare mask */
+		rtccDayCompareModeMonth,    /* Don't care */
+	};
+
+	/* Ensure LE modules are clocked */
+	CMU_ClockEnable(cmuClock_CORELE, true);
+
+#if defined(CMU_LFECLKEN0_RTCC)
+	/* Enable LFECLK in CMU (will also enable oscillator if not enabled) */
+	CMU_ClockSelectSet(cmuClock_LFE, cmuSelect_LFXO);
+#else
+	/* Enable LFACLK in CMU (will also enable oscillator if not enabled) */
+	CMU_ClockSelectSet(cmuClock_LFA, cmuSelect_LFXO);
+#endif
+
+	/* Enable RTCC module clock */
+	CMU_ClockEnable(cmuClock_RTCC, true);
+
+	/* Initialize RTCC */
+	RTCC_Init(&rtcc_config);
+
+	/* Set up compare channels */
+	RTCC_ChannelInit(0, &rtcc_channel_config);
+	RTCC_ChannelInit(1, &rtcc_channel_config);
+	RTCC_ChannelInit(2, &rtcc_channel_config);
+
+	/* Disable module's internal interrupt sources */
+	RTCC_IntDisable(_RTCC_IF_MASK);
+	RTCC_IntClear(_RTCC_IF_MASK);
+
+	/* Clear the counter */
+	RTCC->CNT = 0;
+
+	/* Configure & enable module interrupts */
+	dev_cfg->irq_config();
+
+	LOG_INF("Device %s initialized", DEV_NAME(dev));
+
+	return 0;
+}
+
+static const struct counter_driver_api counter_gecko_driver_api = {
+	.start = counter_gecko_start,
+	.stop = counter_gecko_stop,
+	.read = counter_gecko_read,
+	.set_alarm = counter_gecko_set_alarm,
+	.cancel_alarm = counter_gecko_cancel_alarm,
+	.set_top_value = counter_gecko_set_top_value,
+	.get_pending_int = counter_gecko_get_pending_int,
+	.get_top_value = counter_gecko_get_top_value,
+	.get_max_relative_alarm = counter_gecko_get_max_relative_alarm,
+};
+
+/* RTCC0 */
+
+static struct device DEVICE_NAME_GET(counter_gecko_0);
+
+ISR_DIRECT_DECLARE(counter_gecko_isr_0)
+{
+	struct device *const dev = DEVICE_GET(counter_gecko_0);
+	struct counter_gecko_data *const dev_data = DEV_DATA(dev);
+	counter_alarm_callback_t alarm_callback;
+	u32_t count = RTCC_CounterGet();
+	u32_t flags;
+
+	flags = RTCC_IntGetEnabled();
+	RTCC_IntClear(flags);
+
+	if (flags & RTCC_IF_CC1) {
+		if (dev_data->top_callback) {
+			dev_data->top_callback(dev, dev_data->top_user_data);
+		}
+	}
+	for (int i = 0; i < RTCC_ALARM_NUM; i++) {
+		u8_t cc_idx = chan_id2cc_idx(i);
+
+		if (flags & (RTCC_IF_CC0 << cc_idx)) {
+			if (dev_data->alarm[i].callback) {
+				alarm_callback = dev_data->alarm[i].callback;
+				dev_data->alarm[i].callback = NULL;
+				alarm_callback(dev, i, count,
+					       dev_data->alarm[i].user_data);
+			}
+		}
+	}
+
+	ISR_DIRECT_PM();
+
+	return 1;
+}
+
+#if 0
+BUILD_ASSERT((DT_SILABS_GECKO_RTCC_0_PRESCALER > 0U) &&
+	     (DT_SILABS_GECKO_RTCC_0_PRESCALER <= 32768U));
+#else
+BUILD_ASSERT_MSG(DT_SILABS_GECKO_RTCC_0_PRESCALER == 1U,
+	"In case RTCC prescaler value is different than 1 the module does not "
```

Comment:
Information from an Silabs errata (EFM32PG1B):

> RTCC_E201 – RTCC Does Not Support Compare/Capture Wrap with Prescaler
>
> When the RTCC is configured with a prescaler, the CCV1 top value enable feature enabled by setting CCV1TOP in RTCC_CTRL fails to wrap the counter when RTCC_CNT is equal to RTCC_CC1_CCV, as intended.
>
No workaround no resolution (EFM32PG1B).

For the EFM32PG12B, this issue is not mentioned, but similar ones (RTCC_E205 – Wrap Event Can Be Missed).

Reply:
I can confirm that if i set the prescaler to two:

- On the efr32_slwstk6061a counter_set_top_value() does not work properly. The callback is only called one time. Probably it would be called after a wrap-around of the counter register.

- On the efm32pg_stk3402a counter_set_top_value() works fine. It is called multiple times without calling it again and again.

Reply:
@chrta Thanks a lot for the feedback, very appreciated! 

I've added a run time check to handle errata issue. Please verify.

Reply:
Thanks, the changes look good to me. I can test again in a few days.